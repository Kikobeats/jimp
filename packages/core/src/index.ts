import { Bitmap, Format, JimpClass, Edge } from "@jimp/types";
import { fromBuffer } from "file-type";

const emptyBitmap: Bitmap = {
  data: Buffer.alloc(0),
  width: 0,
  height: 0,
};

export * from "./utils/constants.js";

export interface JimpOptions {
  bitmap?: Bitmap;
}

/** Converts a jimp plugin function to a Jimp class method */
type JimpInstanceMethod<ClassInstance, MethodMap, Method> =
  Method extends JimpMethod<infer Args>
    ? (
        ...args: Args
      ) => JimpInstanceMethods<ClassInstance, MethodMap> & ClassInstance
    : never;

/** Converts a Record of jimp plugin functions to a Record of Jimp class methods */
export type JimpInstanceMethods<ClassInstance, MethodMap> = {
  [Key in keyof MethodMap]: JimpInstanceMethod<
    ClassInstance,
    MethodMap,
    MethodMap[Key]
  >;
};

type JimpMethod<Args extends any[] = any[], J extends JimpClass = JimpClass> = (
  img: J,
  ...args: Args
) => J;

type JimpPlugin = () => { [key: string]: JimpMethod } | void;

type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I,
) => void
  ? I
  : never;

type Constructor<T> = new (...args: any[]) => T;

type JimpFormat<
  M extends string = string,
  T extends Format<M> = Format<M>,
> = () => T;

export function createJimp<
  Methods extends JimpPlugin[],
  Formats extends JimpFormat[],
>({ plugins, formats: formatsArg }: { plugins: Methods; formats?: Formats }) {
  type ExtraMethodMap = JimpInstanceMethods<
    InstanceType<typeof CustomJimp>,
    UnionToIntersection<ReturnType<Methods[number]>>
  >;
  type SupportedMimeTypes = ReturnType<Formats[number]>["mime"];

  const formats = (formatsArg || []).map((format) => format());

  const CustomJimp = class Jimp implements JimpClass {
    /**
     * The bitmap data of the image
     */
    bitmap: Bitmap = emptyBitmap;

    /** Formats that can be used with Jimp */
    formats: Format<any>[] = [];

    constructor(options: JimpOptions = {}) {
      // Add the formats
      this.formats = formats;
      this.bitmap = options.bitmap || emptyBitmap;

      // Add the plugins
      for (const plugin of plugins) {
        const methods = plugin();

        if (methods) {
          for (const key in methods) {
            (this as any)[key] = (...args: any[]) => {
              const result = methods[key]?.(this, ...args);
              if (result) {
                this.bitmap = result.bitmap;
              }
              return this;
            };
          }
        }
      }
    }

    /**
     * Parse a bitmap with the loaded image types.
     *
     * @param {Buffer} data raw image data
     * @param {string} path optional path to file
     * @param {function(Error, Jimp)} finish (optional) a callback for when complete
     * @memberof Jimp
     */
    static async fromBuffer(data: Buffer) {
      const mime = await fromBuffer(data);

      if (!mime || !mime.mime) {
        throw new Error("Could not find MIME for Buffer");
      }

      const format = formats.find((format) => format.mime === mime.mime);

      if (!format || !format.decode) {
        throw new Error(`Mime type ${mime.mime} does not support decoding`);
      }

      return new CustomJimp({
        bitmap: await format.decode(data),
      });
    }

    /**
     * Create a Jimp instance from a bitmap
     * @param bitmap
     * @returns
     */
    static fromBitmap(bitmap: Bitmap) {
      return new CustomJimp({
        bitmap: {
          ...bitmap,
          data: Buffer.from(bitmap.data),
        },
      }) as InstanceType<typeof CustomJimp> & ExtraMethodMap;
    }

    async toBuffer(mime: SupportedMimeTypes) {
      const format = this.formats.find((format) => format.mime === mime);

      if (!format || !format.encode) {
        throw new Error(`Unsupported MIME type: ${mime}`);
      }

      return format.encode(this.bitmap);
    }

    clone() {
      return new CustomJimp({
        bitmap: {
          ...this.bitmap,
          data: Buffer.from(this.bitmap.data),
        },
      });
    }

    /**
     * Returns the offset of a pixel in the bitmap buffer
     * @param x the x coordinate
     * @param y the y coordinate
     * @param edgeHandling (optional) define how to sum pixels from outside the border
     * @returns the index of the pixel or -1 if not found
     */
    getPixelIndex(x: number, y: number, edgeHandling?: Edge) {
      let xi;
      let yi;

      if (!edgeHandling) {
        edgeHandling = Edge.EXTEND;
      }

      if (typeof x !== "number" || typeof y !== "number") {
        throw new Error("x and y must be numbers");
      }

      // round input
      x = Math.round(x);
      y = Math.round(y);
      xi = x;
      yi = y;

      if (edgeHandling === Edge.EXTEND) {
        if (x < 0) xi = 0;
        if (x >= this.bitmap.width) xi = this.bitmap.width - 1;
        if (y < 0) yi = 0;
        if (y >= this.bitmap.height) yi = this.bitmap.height - 1;
      }

      if (edgeHandling === Edge.WRAP) {
        if (x < 0) {
          xi = this.bitmap.width + x;
        }

        if (x >= this.bitmap.width) {
          xi = x % this.bitmap.width;
        }

        if (y < 0) {
          yi = this.bitmap.height + y;
        }

        if (y >= this.bitmap.height) {
          yi = y % this.bitmap.height;
        }
      }

      let i = (this.bitmap.width * yi + xi) << 2;

      // if out of bounds index is -1
      if (xi < 0 || xi >= this.bitmap.width) {
        i = -1;
      }

      if (yi < 0 || yi >= this.bitmap.height) {
        i = -1;
      }

      return i;
    }

    /**
     * Returns the hex color value of a pixel
     * @param x the x coordinate
     * @param y the y coordinate
     * @returns the color of the pixel
     */
    getPixelColor(x: number, y: number) {
      if (typeof x !== "number" || typeof y !== "number") {
        throw new Error("x and y must be numbers");
      }

      const idx = this.getPixelIndex(x, y);
      return this.bitmap.data.readUInt32BE(idx);
    }

    /**
     * Returns the hex colour value of a pixel
     * @param hex color to set
     * @param x the x coordinate
     * @param y the y coordinate
     * @returns the index of the pixel or -1 if not found
     */
    setPixelColor(hex: number, x: number, y: number) {
      if (
        typeof hex !== "number" ||
        typeof x !== "number" ||
        typeof y !== "number"
      ) {
        throw new Error("hex, x and y must be numbers");
      }

      const idx = this.getPixelIndex(x, y);
      this.bitmap.data.writeUInt32BE(hex, idx);

      return this;
    }
  };

  return CustomJimp as typeof CustomJimp & Constructor<ExtraMethodMap>;
}
