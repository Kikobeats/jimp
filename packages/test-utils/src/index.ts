import { expect } from "vitest";
import { Bitmap } from "@jimp/types";
import equal from "fast-deep-equal";

export function hashForEach<Hash extends Record<string, any>>(
  hash: Hash,
  func: (key: keyof Hash, value: Hash[keyof Hash]) => void,
) {
  for (const key in hash)
    if (hash.hasOwnProperty(key)) {
      func(key, hash[key]);
    }
}

function throwUndefinedChar(char: string) {
  const cList: string[] = [];

  hashForEach(colors, (k, c) => {
    cList.push(k + "=" + c.toString(16).padStart(8, "0"));
  });

  throw new Error(
    'The char "' +
      char +
      '" do not defines a color. ' +
      "This are the valid chars: " +
      cList.join(" "),
  );
}

const colors: Record<string | number, number> = {
  "▴": 0xff0000ff, // Red
  "▵": 0xff00007f, // Red half-alpha
  "▸": 0x00ff00ff, // Green
  "▹": 0x00ff007f, // Green half-alpha
  "▾": 0x0000ffff, // Blue
  "▿": 0x0000ff7f, // Blue half-alpha
  "◆": 0xffff00ff, // Yellow
  "◇": 0xffff007f, // Yellow half-alpha
  "▪": 0x00ffffff, // Cyan
  "▫": 0x00ffff7f, // Cyan half-alpha
  "▰": 0xff00ffff, // Magenta
  "▱": 0xff00ff7f, // Magenta half-alpha
  " ": 0x00000000, // Transparent black
  "■": 0x000000ff, // Black
  0: 0x000000ff, // Black
  1: 0x111111ff,
  2: 0x222222ff,
  3: 0x333333ff,
  "▩": 0x404040ff, // Dark gray (1/4 white)
  4: 0x444444ff,
  5: 0x555555ff,
  "!": 0x5f5f5fff,
  6: 0x666666ff,
  7: 0x777777ff,
  8: 0x888888ff,
  "▦": 0x808080ff, // Half gray (1/2 white)
  9: 0x999999ff,
  A: 0xaaaaaaff,
  B: 0xbbbbbbff,
  "▥": 0xbfbfbfff, // Light gray (3/4 white)
  C: 0xccccccff,
  D: 0xddddddff,
  E: 0xeeeeeeff,
  F: 0xffffffff, // White
  "□": 0xffffffff, // White
};

/* Build a test image from a list of strings */
export function makeTestImage(...args: string[]) {
  const initialWidth = args[args.length - 1]?.length || 0;
  const testImage: Bitmap = {
    width: initialWidth,
    height: args.length,
    data: Buffer.alloc(initialWidth * args.length * 4),
  };

  for (let y = 0; y < testImage.height; y++) {
    const line = args[y]!;
    testImage.width = line.length;
    const w = testImage.width;

    for (let x = 0; x < w; x++) {
      const char = line[x]!;
      const color = colors[char];

      if (typeof color === "undefined") {
        throwUndefinedChar(char);
      } else {
        testImage.data.writeUInt32BE(color, 4 * (y * w + x));
      }
    }
  }

  return testImage;
}

const sup = "⁰¹²³⁴⁵⁶⁷⁸⁹ᵃᵇᶜᵈᵉᶠ";

export function testImageReadableMatrix(img: Bitmap) {
  const len = img.data.length / 4;

  const rMatrix = [];
  let line = [];

  for (let i = 0; i < len; i++) {
    const pix = img.data
      .readUint32BE(i * 4)
      .toString(16)
      .toUpperCase()
      .padStart(8, "0");

    line.push(
      pix.replace(/(..)(..)(..)(.)(.)/, (sel, r, g, b, a1, a2) => {
        const a = sup[parseInt(a1, 16)]! + sup[parseInt(a2, 16)]!;
        return r + "-" + g + "-" + b + a;
      }),
    );
    if (i > 0 && (i + 1) % img.width === 0) {
      rMatrix.push(line.join(" "));
      line = [];
    }
  }

  return rMatrix.join("\n");
}

/** Helps to debug image data */
export function testImgToStr(testImage: Bitmap) {
  const colors2: Record<number, string | number> = {};

  hashForEach(colors, (k, c) => {
    colors2[c] = k;
  });

  const lines = [];
  const w = testImage.width;

  for (let y = 0; y < testImage.height; y++) {
    lines[y] = "";
    for (let x = 0; x < w; x++) {
      const cell = testImage.data.readUInt32BE(4 * (y * w + x))!;
      const k = colors2[cell] || "?";
      lines[y] += k;
    }
  }

  return lines.join("\n");
}

function determineBitmapError(testImage: Bitmap, targetTestImage: Bitmap) {
  if (typeof testImage.width === "undefined") {
    return {
      pass: false,
      message: `Expected testImage.width to be defined`,
    };
  }

  if (typeof testImage.height === "undefined") {
    return {
      pass: false,
      message: `Expected testImage.height to be defined`,
    };
  }

  if (testImage.width !== targetTestImage.width) {
    return {
      pass: false,
      message: `Expected testImage.width to be ${targetTestImage.width} but got ${testImage.width}`,
    };
  }

  if (testImage.height !== targetTestImage.height) {
    console.log(typeof testImage.height, typeof targetTestImage.height);
    return {
      pass: false,
      message: `Expected testImage.height to be ${targetTestImage.height} but got ${testImage.height}`,
    };
  }

  if (
    !equal(
      testImageReadableMatrix(testImage),
      testImageReadableMatrix(targetTestImage),
    )
  ) {
    return {
      pass: false,
      message: `Expected testImage:\n${testImgToStr(
        testImage,
      )}\n to be equal to targetTestImage:\n${testImgToStr(targetTestImage)}`,
    };
  }

  return {
    pass: true,
  };
}

export function expectToBeTestImage(
  testImage: Bitmap,
  targetTestImage: Bitmap,
) {
  const error = determineBitmapError(testImage, targetTestImage);

  if (error.pass) {
    return;
  }

  throw new Error(error.message);
}

expect.addSnapshotSerializer({
  serialize(val) {
    const bitmap = "bitmap" in val ? val.bitmap : val;
    return `Visualization:\n\n${testImgToStr(bitmap)}\n\nData:\n\n${testImageReadableMatrix(bitmap)}`;
  },
  test(val) {
    if ("bitmap" in val) {
      return true;
    }

    return val && val.width && val.height && val.data;
  },
});
