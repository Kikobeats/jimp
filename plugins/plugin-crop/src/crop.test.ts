import { expect, test, describe } from "vitest";
import { createJimp } from "@jimp/core";
import { makeTestImage } from "@jimp/test-utils";

import crop from "./index.js";

const jimp = createJimp({ plugins: [crop], formats: [] });

describe("crop", () => {
  // 6x5 size
  const testImage = makeTestImage(
    "  ◆◆  ",
    " ◆▦▦◆ ",
    "◆▦▦▦▦◆",
    " ◆▦▦◆ ",
    "  ◆◆  ",
  );

  test("full width from top", () => {
    expect(jimp.fromBitmap(testImage).crop(0, 0, 6, 2)).toMatchSnapshot();
  });

  test("full width from bottom", () => {
    expect(jimp.fromBitmap(testImage).crop(0, 3, 6, 2)).toMatchSnapshot();
  });

  test("full width from middle", () => {
    expect(jimp.fromBitmap(testImage).crop(0, 2, 6, 2)).toMatchSnapshot();
  });

  test("full height from left", () => {
    expect(jimp.fromBitmap(testImage).crop(0, 0, 2, 5)).toMatchSnapshot();
  });

  test("full height from right", () => {
    expect(jimp.fromBitmap(testImage).crop(4, 0, 2, 5)).toMatchSnapshot();
  });

  test("full height from middle", () => {
    expect(jimp.fromBitmap(testImage).crop(2, 0, 2, 5)).toMatchSnapshot();
  });
});
