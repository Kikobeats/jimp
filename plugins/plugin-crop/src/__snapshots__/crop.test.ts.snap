// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`crop > full height from left 1`] = `
  
 ◆
◆▦
 ◆
  
`;

exports[`crop > full height from middle 1`] = `
◆◆
▦▦
▦▦
▦▦
◆◆
`;

exports[`crop > full height from right 1`] = `
  
◆ 
▦◆
◆ 
  
`;

exports[`crop > full width from bottom 1`] = `
 ◆▦▦◆ 
  ◆◆  
`;

exports[`crop > full width from middle 1`] = `
◆▦▦▦▦◆
 ◆▦▦◆ 
`;

exports[`crop > full width from top 1`] = `
  ◆◆  
 ◆▦▦◆ 
`;
